sql: RBAC telemetry

RichardJCai opened the pull request:
sql: Added anonymized feature tracking for RBAC

Fixes #45022

Release note: None

knz committed:
cli/start: make SIGQUIT dump stacks

What: in the title.

Why: the Go standard behavior is to dump stacks upon SIGQUIT.
This had been disabled since crdb v1.0 but is really useful
to troubleshoot hanging servers.

Release note (cli change): CockroachDB will now dump the stacks of all
goroutines upon receiving SIGQUIT prior to terminating. This feature
is intended for use while troubleshooting misbehaving nodes.

pbardea committed:
storage: re-add MVCC incremental iterator tests

These tests exercise the Go MVCCIncrementalIterator which is no longer
used directly in CockroachDB. However, the MVCCIncrementalIterator is
used as an oracle to test the C++ implementation. Therefore, it would
still be beneficial to have these tests present as long as we plan to
keep these 2 implementations in sync.

Release note: None

petermattis committed:
libroach,cli,server: add thread stack dump facility

Add a facility for dumping the stack traces for all threads in the
process under linux. The technique used was adapted from
github.com/thoughtspot/threadstacks. The list of threads is retrieved by
scanning `/proc/self/tasks`. A realtime signal is sent to each thread
using `rt_tgsigqueueinfo`. A custom signal handler for that signal uses
the glibc `backtrace` facility to retrieve the thread's
stack. Communication between the coordinating thread and the signalled
thread is performed using a pipe (most other synchronization primitives
are not safe to use from a signal handler).

Hook up `/debug/threads` endpoint and add a link on the debug page of
the admin UI. Extend `/_status/stacks` to allow the optional retrieval
of thread stacks (vs the default of goroutine stacks). Enhance `debug
zip` to retrieve the thread stacks for each node.

Release note (admin ui change): Improve debuggability of C++-level
issues by providing access to thread stack traces via a new
`/debug/threads` endpoint which is exposed on the Admin UI advanced
debug page. Include thread stack traces in the info collected by `debug
zip`. Thread stack traces are currently only available on Linux.

a-robinson committed:
cloud/kubernetes: Fix use of deprecated start command in k8s README

And add the useful --logtostderr while I'm here.

pbardea committed:
storageccl: scan all versions of changed key in MVCCIncrementalIterator

This latest implementation of MVCCIncrementalIterator updates its use of
the time-bound iterator to only be used to skip over keys that were not
modified at all within the time bound limits.

Previously, the incremental iterator primarily used the time-bound
iterator and then used a non-time-bound iterator (sanity_iter) to check
that meta keys were properly read. See #43799 for discussion around why
this may not be sufficient.

This commit updates the implementation to use the time-bound iterator
more conservatively. The 2 iterators that it now uses are `iter` (a
non-time-bound iterator and `timeBoundIter`). The main iteration is now
done with `iter`. When `iter` sees a new key, the `timeBoundIter` is
incremented and if it goes further than `iter`, `iter` is `Seek`ed
forward to the meta key of the new key. This means that `iter` should
visit every version of every key that `timeBoundIter` visits.

It should be noted that if many adjacent keys were modified, the above
logic tries to avoid to do a `Seek` to the next key as a `Seek` is ~an
order of magnitude more expensive than a `Next`.

In cases where a lot of keys were modified, we expect to do ~double the
work since 2 iterators would need to seek to every key. However, the
expected use case for the MVCCIncrementalIterator is such that the
_vast_ majority of the data is excluded from the range (as should be the
case for incremental backups), since there is potentially years worth of
data that can be skipped over, versus hours worth of data to include.

Release note: None

RichardJCai committed:
sql: Added anonymized feature tracking for RBAC

Fixes #45022

Release note: None

vladlos committed:
ui: Dashboard

Updated max width dashboard components for match notes signup width

Resolves: #45623

Release note (ui): none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45658)
<!-- Reviewable:end -->

vladlos committed:
ui: Cluster health module

Updated style colors for Cluster health module

Resolves: #45475

Release note (ui): none

nvb committed:
sql/opt/exec: perform implicit SELECT FOR UPDATE on UPDATE with index join

Fixes #45511.

This commit enables implicit row-level locking on UPDATEs with an index-join
in its initial row scan. To support this, two new patterns were needed in
`Builder.shouldApplyImplicitLockingToUpdateInput`:

```
(Update
    $input:(IndexJoin
        $input:(Scan $scanPrivate:*)
        $indexJoinPrivate:*
    )
    $checks:*
    $mutationPrivate:*
)

AND

(Update
    $input:(Project
        $input:(IndexJoin
            $input:(Scan $scanPrivate:*)
            $indexJoinPrivate:*
        )
        $projections:*
        $passthrough:*
    )
    $checks:*
    $mutationPrivate:*
)
```

These extra patterns are pushing the limits of this form of pattern matching,
but this is ok for the next release. If/when we want to make any more general,
we should strongly consider Andy's proposal to fold the patten matching and
analysis into the explorer rather than execbuilder.

tbg committed:
roachtest: skip hotspotsplits on <20.1

Closes #45370.
Closes #45284.
Closes #45045.
Closes #44841.
Closes #44742.

Release note: None

tbg committed:
roachtest: dump stacks when test times out

I've recently seen a number of very opaque test failures in the
acceptance tests, for example `bank/cluster-recovery`. In those
cases there was next to nothing in the logs, and the CRDB logs
looked fine. It will be helpful in this case to have an artifact
in the test directory containing roachtest's stacks, which is
what this commit achieves.

Release note: None

nvb committed:
workload/ycsb: define all fields as NOT NULL

This change marks all columns in the YCSB "usertable" as NOT NULL. Doing
so allows the load generator to take advantage of #44239, which avoids
the KV lookup on the primary column family entirely when querying one of
the other column families. The query plans before and after demonstrate
this:

```
--- Before
root@:26257/ycsb> EXPLAIN SELECT field5 FROM usertable WHERE ycsb_key = 'key';
    tree    |    field    |               description
------------+-------------+------------------------------------------
            | distributed | false
            | vectorized  | false
  render    |             |
   └── scan |             |
            | table       | usertable@primary
            | spans       | /"key"/0-/"key"/1 /"key"/6/1-/"key"/6/2
            | parallel    |

--- After
root@:26257/ycsb> EXPLAIN SELECT field5 FROM usertable WHERE ycsb_key = 'key';
    tree    |    field    |      description
------------+-------------+------------------------
            | distributed | false
            | vectorized  | false
  render    |             |
   └── scan |             |
            | table       | usertable@primary
            | spans       | /"key"/6/1-/"key"/6/2
```

This becomes very important when running YCSB with a column family per
field and with implicit SELECT FOR UPDATE (see #45159). Now that (as
of #45701) UPDATE statements acquire upgrade locks during their initial row
fetch, we don't want them acquiring upgrade locks on the primary column
family of the row they are intending to update a single column in. This
re-introduces the contention between writes to different columns in the
same row that column families helped avoid (see #32704). By marking each
column as NOT NULL, we can continue to avoid this contention.

nvb committed:
workload/ycsb: use SELECT FOR UPDATE in read-modify-write transactions

50% of workload F is read-modify-write transactions. These transactions
scan a single column from a row using a SELECT statement and then modify
that single column using an UPDATE statement. Like most workloads, the
default request distribution is zipfian, meaning that there are some rows
that are heavily contended.

This is the perfect use case for explicit SELECT FOR UPDATE. Using it for
the "read" portion of the read-modify-write transaction dramatically improves
throughput and eliminates transaction retries. This commit makes that change.

Confusingly, these are not implemented as real transactions in the official
YCSB load generator, but that seems strictly incorrect and was likely
dictated by the lack of support for transactions in NoSQL engines of the
time:
  https://github.com/brianfrankcooper/YCSB/blob/cd1589ce6f5abf96e17aa8ab80c78a4348fdf29a/core/src/main/java/site/ycsb/workloads/CoreWorkload.java#L729

Also, strangely Workload F wasn't in the initial YCSB paper:
  https://www2.cs.duke.edu/courses/fall13/compsci590.4/838-CloudPapers/ycsb.pdf
But it was in the initial commit of the YCSB load generator:
  https://github.com/brianfrankcooper/YCSB/commit/24efaff35cb20b4ae730b3f23716ed73fea783d9#diff-d240bdb33fb042b95010733db8ccec63

This might break compatibility with old versions of CRDB, as I think we
only started accepting SELECT FOR UPDATE in 19.2. The commit includes a
flag to avoid the functionality, if necessary.

sumeerbhola committed:
kvserver/concurrency: misc lock table tests to cover functionality
that was not tested before and three bug fixes

Some of the following was based on looking at test coverage results:
- added wait_self: covers reservation by a different request from the
  the same txn and lock acquisition by the same txn
- we were not adding to the sequence nums in discoveredLock
  which was a bug found when printing the sequence numbers
- dup_access: added test when a request is both reading and writing
  to a key and the lock is held after it stops waiting for write
- added lock_changes: misc state changes for a lock that were not covered
  by existing tests
- added ignored seqnum tests to the update testdata file. It found
  a bug
- UpdateLocks now ignores updates with the same epoch but lower timestamp.
  It used to trigger an error after the ts had been updated which was
  severely muddled (tell the caller of the error after making the erroneous
  state change!).

Release note: None

sumeerbhola committed:
kvserver/concurrency: changed behavior of lock update

Needs a review to check that this is the desired behavior,
after which I will add a test.

Release note: None

pbardea committed:
importccl: handle protected timestamp removal upon resume

If a protected timestamp is removed and the import job is resumed, we
should still be able to mark the job as successfully complete.

Release note: None

pbardea committed:
sql: allow single-char usernames

Previously, the regex which validated usernames changed from
`^[\p{Ll}0-9_][\p{Ll}0-9_-]{0,62}$` to `^[\p{Ll}0-9_][---\p{Ll}0-9_.]+$`
however in the process also prevented usernames of length 1 to be
created (we previously supported this, and so does Postgres). This
commit changes the + to a *.

Release note: None

knz committed:
cli: report partial result rows in case of error

Before this patch:

```
root@127.0.0.1:61756/defaultdb> select 1/(@1-10) from generate_series(1,11);
ERROR: division by zero
SQLSTATE: 22012
```

After:

```
root@127.0.0.1:61756/defaultdb> select 1/(@1-10) from generate_series(1,11);
         ?column?
---------------------------
  -0.11111111111111111111
                   -0.125
  -0.14285714285714285714
  -0.16666666666666666667
                     -0.2
                    -0.25
  -0.33333333333333333333
                     -0.5
                       -1
(9 rows)
(error encountered after some results rows were delivered)
ERROR: division by zero
SQLSTATE: 22012
```

Release note (cli change): In case some result rows have already been
received from the server when an error is encountered, CockroachDB's
SQL shell (`cockroach sql/demo`) now present both the result rows and
the error in the output regardless of the selected table
formatter. Previously, only the error was reported with some
formatters, or both with other formatters.

tbg committed:
bulk: git mv to ./pkg/kv

Pure code movement, does not compile.

Release note: None

tbg committed:
bulk: Goland-move to ./pkg/kv/bulk

Release note: None

tbg committed:
cloud: git mv to ./pkg/storage/cloud

Release note: None

tbg committed:
cloud: Goland-move to ./pkg/storage/cloud

Release note: None

tbg committed:
kv: git mv move package content to kvclient/kvcoord

Release note: None

tbg committed:
kv: Goland-move to kvclient/kvcoord

Release note: None

tbg committed:
kv,kvclient: add placeholder files

TestForbiddenDeps trips over these empty directories, so make them
trivial Go packages. They will fill with live shortly anyway.

Release note: None

tbg committed:
internal/client: git mv *.go to kv package

Release note: None

tbg committed:
internal/client: Goland-move to pkg/kv

Release note: None

tbg committed:
kv: remove placeholder file

Release note: None

tbg committed:
bulk: fix an import path

Not amending the commit since it already skews.

Release note: None

tbg committed:
github-post: fix TestListFailures

Test needed an updated package name.

Release note: None

tbg committed:
kv: fix various go generate invocations

Needed to adjust relative paths.

Release note: None

tbg committed:
cloud,kv,kvcoord: fix lint

Release note: None

sumeerbhola committed:
kvserver/concurrency: changed behavior of lock update and addressed
comments

Release note: None

RichardJCai committed:
sql: allow roles to login, add tests for logging into roles

Had a logic error in checking whether or not roles could login.
Fixing the logic check to correctly allow roles to login.

Release note: none

RichardJCai commented:
Messed up the commit history after a rebase -- gonna remake PR to make it easier to navigate
